.github/workflows/test.yaml

```diff
@@ -16,4 +28,5 @@ jobs:
           DOCKER_DEV_CI_MODE: 1
           DOCKER_BUILDKIT: 0
           DOCKER_QUIET_BUILD: 1
-        run: make -C utils/docker run DOCKER_DEV_OPTIONAL_CMD='env DOCKER_QUIET_BUILD=1 DOCKER_BUILDKIT=0 make all clobber'
\ No newline at end of file
+        run: |
+            DOCKER_BUILD_OPTS="--build-arg UBUNTU_VERSION=${{ matrix.os-version }} --build-arg UBUNTU_NAME=${{ matrix.os-name }}" make -C utils/docker run DOCKER_DEV_OPTIONAL_CMD='env DOCKER_QUIET_BUILD=1 DOCKER_BUILDKIT=0 make all clobber'
```

Comment:
Hmm, i think you have to add the DOCKER_BUILD_OPTS also to the second `env` clause of the OS choice will not be passed transitively to the docker-builds inside container. The dev container definitely should be built in correct OS but without defining the DOCKER_BUILD_OPTS also inside the container itself the build inside container would always build define 20.04 ?

Reply:
This is incredibly confusing. Let's try to clarify a few things.

The `make run` has some dependencies (on `dev`, `base-dev`, etc.), which build images and pass the `DOCKER_BUILD_OPTS` var. So, the built images should be as we intend them to be.

Then, `make run` enters the container to run what is in `DOCKER_DEV_OPTIONAL_CMD` -- essentially the `make all`.
Most importantly, it does not propagate `DOCKER_BUILD_OPTS`.
Question: does this trigger the (re-)build of the images? (If not, there's nothing to do.)

If so, the dockerfile will use the default value (i.e. 20.04 focal), and that's not intended.
Two solutions:
1. as suggested, pass `DOCKER_BUILD_OPTS` as part of env
2. let `make run` propagate `DOCKER_BUILD_OPTS`
Question: which one is more appropriate? (I'd argue the second one)

Comments?

Reply:
> This is incredibly confusing. Let's try to clarify a few things.
> 
> The `make run` has some dependencies (on `dev`, `base-dev`, etc.), which build images and pass the `DOCKER_BUILD_OPTS` var. So, the built images should be as we intend them to be.

yes
>
> Then, `make run` enters the container to run what is in `DOCKER_DEV_OPTIONAL_CMD` -- essentially the `make all`.

yes (and it does so with the default environement as defined by default shell and (default) `config.mk` as well as the explicitly defined variables using the `env` prefix ...)

> Most importantly, it does not propagate `DOCKER_BUILD_OPTS`.

Yes as the docker container does not inherit the environment of the process calling the `make run`, i.e., any variable you want to inherit you have to explicitly pass using the `env` (which previously we did for the docker quiet and build options, but now you also would have to do for DOCKER_BUILD_OPTS. An alternative might be pass them also via docker-mounted `config.override.mk` or alike)

> Question: does this trigger the (re-)build of the images? (If not, there's nothing to do.)

`make` _always_ causes a `docker build` (it is nearly impossible in make to add dependencies which would figure out whether the build really can be skipped ...) but docker via caching tries to skip layers with dependencies which have not changed.   But also note that the `make run` from the github action only builds the `dev` container whereas `make all` builds a whole bunch of other ones ...

> If so, the dockerfile will use the default value (i.e. 20.04 focal), and that's not intended.
> Two solutions:
> 
> 1. as suggested, pass `DOCKER_BUILD_OPTS` as part of env
> 2. let `make run` propagate `DOCKER_BUILD_OPTS`
>    Question: which one is more appropriate? (I'd argue the second one)

Both are options, but we should do it uniformly also for the other options/flags and not only for `DOCKER_BUILD_OPTS`. So (1) in a way is the easier option while (2) is the cleaner but requires a bit more thinking what all has to be passed  ....

> Comments?

Reply:
In the interest of time, the easier is preferred. Now we should be good.